### PR TITLE
Cli/setup

### PR DESCRIPTION
**Pull request by bytemare**
Added command for PKI setup.

`"./netm4ul pki [options]"` creates a CA, a server and a specified amount of client certificates signed with the CA compatible for mutual TLS authentication. 

TODO : 
- deploy certificates to server and nodes through Ansible
- adapt server and client to point to correct files (minor difficulty)

### CODE REVIEW

**Comment by Edznux** [comment]
PR against develop plz

File: cmd/pki.go [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+	// Define timespan the certificate has to be valid for
+	notBefore := time.Now()
+	notAfter := notBefore.Add(validFor)
+
+	// Generate a random serial number
+	serialNumber := generateSerial()
+
+	// Define a name for the certificate
+	certFilename := destDir + "/" + entityID + "/" + entityID + "_cert.pem"
+	keyFilename := destDir + "/" + entityID + "/" + entityID + "_key.pem"
```
Comment (Edznux) [inline cmd/pki.go]: Use Join() instead

File: cmd/pki.go [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+	}
+
+	return nil, nil, certFilename, keyFilename
+}
+
+// Checks if directory exists. If not, creates it.
+func checkDir(targetDirectory string) {
+
+	//newpath := filepath.Join(".", targetDirectory)
+	err := os.MkdirAll(targetDirectory, 0700)
```
Comment (Edznux) [inline cmd/pki.go]: _MkdirAll creates a directory named path, along with any necessary parents, and returns nil, or else returns an error. The permission bits perm (before umask) are used for all directories that MkdirAll creates. If path is already a directory, MkdirAll does nothing and returns nil._

It seems that you're not handling any errors there.

File: cmd/pki.go [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+
+	pkiCmd.Flags().UintVarP(&nNodes, "nodes", "n", 1, "Number of nodes to create certificates for")
+	pkiCmd.Flags().StringVarP(&serverID, "server", "@", "", "Server address, with IP or DNS name")
+	pkiCmd.Flags().StringVarP(&ecdsaCurve, "ec", "e", "P256", "Elliptic curve to use for certificates. Accepted values : P224, P256 (recommended), P384 and P521")
+	pkiCmd.Flags().BoolVarP(&keepPrivCA, "keepca", "k", false, "Indicate whether to keep CA private key (i.e. written to disk)")
+	pkiCmd.Flags().DurationVarP(&certDuration, "duration", "d", 365*24*time.Hour, "Duration for certificate. Default is a year")
+	pkiCmd.Flags().StringVarP(&organisationSubject, "org", "o", "Netm4ul", "Organisation Subject to use in server certificate")
+	pkiCmd.Flags().StringVarP(&certificateDirectory, "dir", "t", "./certificates", "Local directory to store PKI certs and keys")
+
+	viper.SetDefault("ec", "P256")
```
Comment (Edznux) [inline cmd/pki.go]: Is this really required ?
Default value are defined just above or am I missing something?

File: cmd/pki.go [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+
+	nowH, nowM, nowS := now.Clock()
+
+	return fmt.Sprintf("%s %d %d:%d:%d %d", now.Month(), now.Day(), nowH, nowM, nowS, now.Year())
+}
+*/
+
+// Generates a ecdsa private key from a cryptographically secure prng
+func generatePrivateKey(ecdsaCurve string) interface{} {
+	var priv interface{}
```
Comment (Edznux) [inline cmd/pki.go]: Is the `interface{}` type required here ?
Maybe [ecdsa.PrivateKey](https://golang.org/pkg/crypto/ecdsa/#PrivateKey) could be used?

File: cmd/pki.go [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+	},
+
+	Run: func(cmd *cobra.Command, args []string) {
+		fmt.Println("Setting up PKI at " + certificateDirectory)
+		pkiSetup(organisationSubject, nNodes, certDuration, ecdsaCurve, certificateDirectory, serverID)
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(pkiCmd)
```
Comment (Edznux) [inline cmd/pki.go]: I think we should not add this command to the `rootCmd` but on the `setupCmd`.
This is part of the setup phase.

File: cmd/pki.go [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+	caCert, caKey, _, _ := create("ca", certDuration, "CA", ecdsaCurve, nil, nil, organisationSubject, certificateDirectory)
+	if caCert == nil || caKey == nil {
+		log.Print("Failed to created CA cert\n")
+	}
+
+	// Create Server
+	_, _, _, _ = create(serverID, certDuration, "Server", ecdsaCurve, caCert, caKey, organisationSubject, certificateDirectory)
+
+	// Create clients
+	clientsDir := certificateDirectory + "/clients"
```
Comment (Edznux) [inline cmd/pki.go]: Use Join()

File: cmd/pki.go [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+		pem.Encode(keyOut, pemBlockForKey(privateKey))
+		keyOut.Close()
+		// log.Print("wrote " + keyFilename + "\n")
+		// log.Print("Successfully achieved job for " + keyFilename + "\n")
+	}
+
+}
+
+// Creates the certificate for the given entity and writes key material to disk. In case of the CA, also return the certificate and private key to sign the end entity parties
+func create(entityID string, validFor time.Duration, entityType string, ecdsaCurve string, signerCert *x509.Certificate, signerPrivKey interface{}, organisation string, destDir string) (*x509.Certificate, interface{}, string, string) {
```
Comment (Edznux) [inline cmd/pki.go]: Can you split this function ?
Line 337, you are passing 2 nil parameter and use only 2 return value
Line 343 and 348 should call a simpler dedicated function ?
If this is not a CA (line 318) you always return 2 nil value
`createServer`, `createClient` and `createCA` maybe ?

**Review by Edznux — CHANGES_REQUESTED** [review]
Cool, thanks.
There are some required changes in the comments.
And I not sure about something : its seems that half of an RSA support is implemented. Should it be removed first ?
Line 103 is commented out, but 114 is not.
the `generatePrivateKey` returns only ecdsa keys (but use and `interface{}` type.

If the RSA implementation is removed, please change all `interface{}` to ecdsa.PrivateKey.

File: cmd/pki.go (hunk 1 above)
Comment (bytemare) [reply cmd/pki.go]: Huh, good point.

File: cmd/pki.go (hunk 2 above)
Comment (bytemare) [reply cmd/pki.go]: Actually, this only served me to check if the dir existed and pop a warning. Let me do it in a better way.

File: cmd/pki.go (hunk 3 above)
Comment (bytemare) [reply cmd/pki.go]: Wasn't sure how to play with Cobra and Viper. Actually, using Viper displays "(default [VALUE]") on help command. I should actually remove values from Flags() lines.

File: cmd/pki.go (hunk 4 above)
Comment (bytemare) [reply cmd/pki.go]: Like discussed about possible but undesired RSA support, I will have a look. Let's toss the support and go full ec.

File: cmd/pki.go (hunk 5 above)
Comment (bytemare) [reply cmd/pki.go]: ok. I wasn't sure. Will do.

File: cmd/pki.go (hunk 6 above)
Comment (bytemare) [reply cmd/pki.go]: +1

File: cmd/pki.go (hunk 7 above)
Comment (bytemare) [reply cmd/pki.go]: You're right. I kept the structure of my POC. Splitting will clear the code.

**Review by Edznux — APPROVED** [review]

**Pull request merged**